HDR 2D Canvas: Support for Float16 backend with put/getImageData

squelart opened the pull request:
#### 7a7acedd6de93309eae61d40394f0fbbb886541e
<pre>
HDR 2D Canvas: Support for Float16 backend with put/getImageData
<a href="https://bugs.webkit.org/show_bug.cgi?id=297212">https://bugs.webkit.org/show_bug.cgi?id=297212</a>
<a href="https://rdar.apple.com/158032286">rdar://158032286</a>

Reviewed by Mike Wyrzykowski.

ImageBufferBackend&apos;s support for Float16 image data, including
conversions between uint8 and float16.

In this initial &quot;testable&quot; implementation, setting the canvas pixel
format to float16 is enough to engage the display of HDR pixels.

* LayoutTests/TestExpectations:
* LayoutTests/fast/canvas/hdr/float16-canvas-imagedata-expected.txt: Added.
* LayoutTests/fast/canvas/hdr/float16-canvas-imagedata.html: Added.
* LayoutTests/fast/canvas/imagedata-storageformat-enabled.html:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/storage/indexeddb/modern/objectstore-autoincrement-types-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
(WebCore::ImageData::float16ArrayPixelBuffer const):
(WebCore::ImageData::pixelBuffer const):
* Source/WebCore/html/ImageData.h:
(WebCore::ImageData::create):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::cacheImageDataIfPossible):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
(WebCore::CanvasRenderingContext2DBase::putImageData):
* Source/WebCore/platform/graphics/ImageBufferAllocator.cpp:
(WebCore::ImageBufferAllocator::createPixelBuffer const):
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::getPixelBuffer):
(WebCore::ImageBufferBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/PixelBuffer.cpp:
(WebCore::computeRawPixelComponentCount):
(WebCore::PixelBuffer::computeBufferSize):
(WebCore::PixelBuffer::PixelBuffer):
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBuffer::bytesPerPixelComponent):
(WebCore::PixelBuffer::componentsPerPixel):
(WebCore::PixelBuffer::bytesPerPixel):
* Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
(WebCore::readFloat16):
(WebCore::writeFloat16):
(WebCore::convertImagePixelsFromFloat16ToFloat16):
(WebCore::convertImagePixelsFromFloat16):
(WebCore::convertImagePixelsToFloat16):
(WebCore::convertImagePixels):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.cpp:
(WebCore::ImageBufferCGBackend::calculateBytesPerRow):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::calculateMemoryCost):
(WebCore::ImageBufferCGBitmapBackend::bytesPerRow const):
* Source/WebCore/platform/graphics/cg/ImageBufferCGPDFDocumentBackend.cpp:
(WebCore::ImageBufferCGPDFDocumentBackend::calculateMemoryCost):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::calculateBytesPerRow):
(WebCore::ImageBufferIOSurfaceBackend::calculateMemoryCost):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingForCanvas const):
* Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.cpp:
(WebKit::ShareablePixelBuffer::ShareablePixelBuffer):
(WebKit::ShareablePixelBuffer::createScratchPixelBuffer const):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::bytesPerRow const):

Canonical link: <a href="https://commits.webkit.org/298618@main">https://commits.webkit.org/298618@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d44b8891fad44e7c8ea398890d9375a42a8cba32

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116057 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35718 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26260 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122113 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66606 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8391745a-a7e3-4673-ac9a-77b6af23178e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117946 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36412 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44306 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88168 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f1cc40df-3b64-4510-bec3-b346dbe6fd71) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119005 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29045 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104140 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68579 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28163 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22248 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65795 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108166 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98433 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22385 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125263 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114585 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42951 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32241 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96905 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43316 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100330 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96689 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24605 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41961 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19844 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42838 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48430 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143269 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42305 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36934 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45640 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44009 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
```diff
@@ -316,9 +318,179 @@ static void copyImagePixels(const ConstPixelBufferConversionView& source, const
+    union {
+        Float16 float16 { };
+        uint8_t bytes[sizeof(Float16)];
+    } magicBox;
```
magicBox seems like an interesting name but it’s a local union so that seems fine

squelart replied:
You're right, it was a bit too whimsical, and showing my distrust of technically-UB conversions through unions.
I'll change it to a more descriptive "float/pixel16OrBytesUnion".

squelart committed:
HDR 2D Canvas: Support for Float16 backend with put/getImageData
https://bugs.webkit.org/show_bug.cgi?id=297212
rdar://158032286

Reviewed by Mike Wyrzykowski.

ImageBufferBackend's support for Float16 image data, including
conversions between uint8 and float16.

In this initial "testable" implementation, setting the canvas pixel
format to float16 is enough to engage the display of HDR pixels.

* LayoutTests/TestExpectations:
* LayoutTests/fast/canvas/hdr/float16-canvas-imagedata-expected.txt: Added.
* LayoutTests/fast/canvas/hdr/float16-canvas-imagedata.html: Added.
* LayoutTests/fast/canvas/imagedata-storageformat-enabled.html:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/storage/indexeddb/modern/objectstore-autoincrement-types-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
(WebCore::ImageData::float16ArrayPixelBuffer const):
(WebCore::ImageData::pixelBuffer const):
* Source/WebCore/html/ImageData.h:
(WebCore::ImageData::create):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::cacheImageDataIfPossible):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
(WebCore::CanvasRenderingContext2DBase::putImageData):
* Source/WebCore/platform/graphics/ImageBufferAllocator.cpp:
(WebCore::ImageBufferAllocator::createPixelBuffer const):
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::getPixelBuffer):
(WebCore::ImageBufferBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/PixelBuffer.cpp:
(WebCore::computeRawPixelComponentCount):
(WebCore::PixelBuffer::computeBufferSize):
(WebCore::PixelBuffer::PixelBuffer):
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBuffer::bytesPerPixelComponent):
(WebCore::PixelBuffer::componentsPerPixel):
(WebCore::PixelBuffer::bytesPerPixel):
* Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
(WebCore::readFloat16):
(WebCore::writeFloat16):
(WebCore::convertImagePixelsFromFloat16ToFloat16):
(WebCore::convertImagePixelsFromFloat16):
(WebCore::convertImagePixelsToFloat16):
(WebCore::convertImagePixels):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.cpp:
(WebCore::ImageBufferCGBackend::calculateBytesPerRow):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::calculateMemoryCost):
(WebCore::ImageBufferCGBitmapBackend::bytesPerRow const):
* Source/WebCore/platform/graphics/cg/ImageBufferCGPDFDocumentBackend.cpp:
(WebCore::ImageBufferCGPDFDocumentBackend::calculateMemoryCost):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::calculateBytesPerRow):
(WebCore::ImageBufferIOSurfaceBackend::calculateMemoryCost):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingForCanvas const):
* Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.cpp:
(WebKit::ShareablePixelBuffer::ShareablePixelBuffer):
(WebKit::ShareablePixelBuffer::createScratchPixelBuffer const):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::bytesPerRow const):

Canonical link: https://commits.webkit.org/298618@main
